server/src/main/java/org/elasticsearch/index/mapper/TermBasedFieldType.java

```diff
@@ -55,15 +60,33 @@ public boolean mayExistInIndex(SearchExecutionContext context) {
 
     @Override
     public Query termQuery(Object value, SearchExecutionContext context) {
-        failIfNotIndexed();
-        return new TermQuery(new Term(name(), indexedValueForSearch(value)));
+        if (allowDocValueBasedQueries()) {
+            failIfNotIndexedNorDocValuesFallback(context);
+        } else {
+            failIfNotIndexed();
+        }
+        if (isIndexed()) {
+            return new TermQuery(new Term(name(), indexedValueForSearch(value)));
+        } else {
+            // Should we use SortedSetDocValuesField.newSlowExactQuery instead?
+            return new DocValuesTermsQuery(name(), indexedValueForSearch(value));
+        }
     }
 
     @Override
     public Query termsQuery(Collection<?> values, SearchExecutionContext context) {
-        failIfNotIndexed();
+        if (allowDocValueBasedQueries()) {
+            failIfNotIndexedNorDocValuesFallback(context);
+        } else {
+            failIfNotIndexed();
+        }
         BytesRef[] bytesRefs = values.stream().map(this::indexedValueForSearch).toArray(BytesRef[]::new);
-        return new TermInSetQuery(name(), bytesRefs);
+        if (isIndexed()) {
+            return new TermInSetQuery(name(), bytesRefs);
+        } else {
+            // Should we use super.termsQuery instead?
+            return new DocValuesTermsQuery(name(), bytesRefs);
```

Comment:
@jpountz thoughts on which query to use?

Reply:
+1 to DocValuesTermsQuery

Reply:
👍  thanks for confirming.